File: src/services/badgePollingService.ts
// Badge Polling Service
// Polls for new badges and emits notifications when new badges are detected
// or when badges become available

import { invoke } from '@tauri-apps/api/core';

const POLL_INTERVAL_MS = 5 * 60 * 1000; // 5 minutes
const KNOWN_BADGES_KEY = 'streamnook_known_badges';
const NOTIFIED_AVAILABLE_KEY = 'streamnook_notified_available_badges';

export interface BadgeVersion {
    id: string;
    image_url_1x: string;
    image_url_2x: string;
    image_url_4x: string;
    title: string;
    description: string;
}

export interface BadgeMetadata {
    date_added: string | null;
    usage_stats: string | null;
    more_info: string | null;
    info_url: string;
}

export interface BadgeSet {
    set_id: string;
    versions: BadgeVersion[];
}

export interface BadgeWithStatus {
    set_id: string;
    version: BadgeVersion;
    status: 'available' | 'coming_soon' | 'expired' | null;
    metadata?: BadgeMetadata;
}

export interface BadgeNotification {
    badge_name: string;
    badge_set_id: string;
    badge_version: string;
    badge_image_url: string;
    badge_description?: string;
    status: 'new' | 'available' | 'coming_soon';
    date_info?: string;
}

type BadgeEventCallback = (badges: BadgeNotification[]) => void;

class BadgePollingService {
    private intervalId: ReturnType<typeof setInterval> | null = null;
    private isPolling = false;
    private listeners: BadgeEventCallback[] = [];
    private lastPollTimestamp: number = 0;

    // Start polling for badges
    start() {
        if (this.intervalId) {
            console.log('[BadgePolling] Already running');
            return;
        }

        console.log('[BadgePolling] Starting badge polling service');

        // Do an initial poll after a short delay to let the app settle
        setTimeout(() => {
            this.poll();
        }, 10000); // 10 seconds

        // Then poll every 5 minutes
        this.intervalId = setInterval(() => {
            this.poll();
        }, POLL_INTERVAL_MS);
    }

    // Stop polling
    stop() {
        if (this.intervalId) {
            clearInterval(this.intervalId);
            this.intervalId = null;
            console.log('[BadgePolling] Stopped badge polling service');
        }
    }

    // Subscribe to badge notifications
    subscribe(callback: BadgeEventCallback) {
        this.listeners.push(callback);
        return () => {
            this.listeners = this.listeners.filter(l => l !== callback);
        };
    }

    // Emit badge notifications to all listeners
    private emit(badges: BadgeNotification[]) {
        this.listeners.forEach(listener => listener(badges));
    }

    // Get known badges from localStorage
    private getKnownBadges(): Set<string> {
        try {
            const stored = localStorage.getItem(KNOWN_BADGES_KEY);
            if (stored) {
                return new Set(JSON.parse(stored));
            }
        } catch {
            // Ignore parse errors
        }
        return new Set();
    }

    // Save known badges to localStorage
    private saveKnownBadges(badges: Set<string>) {
        try {
            localStorage.setItem(KNOWN_BADGES_KEY, JSON.stringify([...badges]));
        } catch (error) {
            console.warn('[BadgePolling] Failed to save known badges:', error);
        }
    }

    // Get badges we've already notified about being available
    private getNotifiedAvailable(): Set<string> {
        try {
            const stored = localStorage.getItem(NOTIFIED_AVAILABLE_KEY);
            if (stored) {
                return new Set(JSON.parse(stored));
            }
        } catch {
            // Ignore parse errors
        }
        return new Set();
    }

    // Save notified available badges to localStorage
    private saveNotifiedAvailable(badges: Set<string>) {
        try {
            localStorage.setItem(NOTIFIED_AVAILABLE_KEY, JSON.stringify([...badges]));
        } catch (error) {
            console.warn('[BadgePolling] Failed to save notified available badges:', error);
        }
    }

    // Parse date range from more_info field
    private parseDateRange(text: string): { start: Date; end: Date } | null {
        const months: Record<string, number> = {
            'Jan': 0, 'Feb': 1, 'Mar': 2, 'Apr': 3,
            'May': 4, 'Jun': 5, 'Jul': 6, 'Aug': 7,
            'Sep': 8, 'Oct': 9, 'Nov': 10, 'Dec': 11
        };
        const currentYear = new Date().getFullYear();

        // Match "Mon DD – Mon DD" format (e.g., "Dec 06 – Dec 07")
        const fullRangeMatch = text.match(/(\w{3})\s+(\d{1,2})\s*[–-]\s*(\w{3})\s+(\d{1,2})/);
        if (fullRangeMatch) {
            const startMonthAbbrev = fullRangeMatch[1];
            const startDay = parseInt(fullRangeMatch[2], 10);
            const endMonthAbbrev = fullRangeMatch[3];
            const endDay = parseInt(fullRangeMatch[4], 10);

            if (months.hasOwnProperty(startMonthAbbrev) && months.hasOwnProperty(endMonthAbbrev)) {
                const startDate = new Date(currentYear, months[startMonthAbbrev], startDay, 0, 0, 0);
                const endDate = new Date(currentYear, months[endMonthAbbrev], endDay, 23, 59, 59);
                return { start: startDate, end: endDate };
            }
        }

        // Match "Mon D-D" or "Mon D - D" format (e.g., "Dec 1-12" or "Dec 1 - 12")
        const shortRangeMatch = text.match(/(\w{3})\s+(\d{1,2})\s*[–-]\s*(\d{1,2})(?!\s*\w)/);
        if (shortRangeMatch) {
            const monthAbbrev = shortRangeMatch[1];
            const startDay = parseInt(shortRangeMatch[2], 10);
            const endDay = parseInt(shortRangeMatch[3], 10);

            if (months.hasOwnProperty(monthAbbrev)) {
                const monthNum = months[monthAbbrev];
                const startDate = new Date(currentYear, monthNum, startDay, 0, 0, 0);
                const endDate = new Date(currentYear, monthNum, endDay, 23, 59, 59);
                return { start: startDate, end: endDate };
            }
        }

        return null;
    }

    // Get badge status based on more_info metadata
    private getBadgeStatus(metadata?: BadgeMetadata): { status: 'available' | 'coming_soon' | 'expired' | null; dateInfo?: string } {
        const moreInfo = metadata?.more_info;
        if (!moreInfo) return { status: null };

        const now = Date.now();
        const dateRange = this.parseDateRange(moreInfo);

        if (dateRange) {
            const startTime = dateRange.start.getTime();
            const endTime = dateRange.end.getTime();

            // Extract date info for display
            const dateMatch = moreInfo.match(/(\w{3}\s+\d{1,2}(?:\s*[–-]\s*(?:\w{3}\s+)?\d{1,2})?)/);
            const dateInfo = dateMatch ? dateMatch[1] : undefined;

            if (now < startTime) {
                return { status: 'coming_soon', dateInfo };
            } else if (now >= startTime && now <= endTime) {
                return { status: 'available', dateInfo };
            } else {
                return { status: 'expired', dateInfo };
            }
        }

        return { status: null };
    }

    // Main poll function
    async poll() {
        if (this.isPolling) {
            console.log('[BadgePolling] Already polling, skipping');
            return;
        }

        this.isPolling = true;
        console.log('[BadgePolling] Polling for badge updates...');

        try {
            // Get cached badges first
            const cachedBadges = await invoke<{ data: BadgeSet[] } | null>('get_cached_global_badges');

            if (!cachedBadges || !cachedBadges.data || cachedBadges.data.length === 0) {
                console.log('[BadgePolling] No cached badges, skipping poll');
                this.isPolling = false;
                return;
            }

            // Get known badges and notified available badges
            const knownBadges = this.getKnownBadges();
            const notifiedAvailable = this.getNotifiedAvailable();
            const notifications: BadgeNotification[] = [];

            // Flatten badges
            const allBadges: { set_id: string; version: BadgeVersion }[] = [];
            for (const set of cachedBadges.data) {
                for (const version of set.versions) {
                    allBadges.push({ set_id: set.set_id, version });
                }
            }

            // Check each badge
            for (const { set_id, version } of allBadges) {
                const badgeKey = `${set_id}-v${version.id}`;

                // Try to get metadata for this badge
                let metadata: BadgeMetadata | undefined;
                try {
                    const cached = await invoke<{ data: BadgeMetadata } | null>('get_universal_cached_item', {
                        cacheType: 'badge',
                        id: `metadata:${badgeKey}`,
                    });
                    if (cached) {
                        metadata = cached.data;
                    }
                } catch {
                    // Metadata not available, that's okay
                }

                // Get current status
                const { status, dateInfo } = this.getBadgeStatus(metadata);

                // Check if this is a new badge we haven't seen before
                if (!knownBadges.has(badgeKey)) {
                    console.log(`[BadgePolling] New badge detected: ${version.title} (${badgeKey})`);

                    // Only notify for new badges if they're available or coming soon
                    if (status === 'available' || status === 'coming_soon') {
                        notifications.push({
                            badge_name: version.title,
                            badge_set_id: set_id,
                            badge_version: version.id,
                            badge_image_url: version.image_url_4x || version.image_url_2x || version.image_url_1x,
                            badge_description: version.description,
                            status: 'new',
                            date_info: dateInfo,
                        });
                    } else if (!status) {
                        // If no time-sensitive status, still notify as new
                        notifications.push({
                            badge_name: version.title,
                            badge_set_id: set_id,
                            badge_version: version.id,
                            badge_image_url: version.image_url_4x || version.image_url_2x || version.image_url_1x,
                            badge_description: version.description,
                            status: 'new',
                        });
                    }

                    // Mark as known
                    knownBadges.add(badgeKey);
                }

                // Check if a known badge just became available (status changed)
                if (knownBadges.has(badgeKey) && status === 'available' && !notifiedAvailable.has(badgeKey)) {
                    console.log(`[BadgePolling] Badge now available: ${version.title} (${badgeKey})`);

                    notifications.push({
                        badge_name: version.title,
                        badge_set_id: set_id,
                        badge_version: version.id,
                        badge_image_url: version.image_url_4x || version.image_url_2x || version.image_url_1x,
                        badge_description: version.description,
                        status: 'available',
                        date_info: dateInfo,
                    });

                    // Mark as notified for availability
                    notifiedAvailable.add(badgeKey);
                }
            }

            // Save updated known badges and notified available
            this.saveKnownBadges(knownBadges);
            this.saveNotifiedAvailable(notifiedAvailable);

            // Emit only the most recent/latest badge notification (not the entire list)
            if (notifications.length > 0) {
                // Take only the most recent badge (last one added)
                const latestBadge = notifications[notifications.length - 1];
                console.log(`[BadgePolling] Emitting 1 badge notification (${latestBadge.badge_name}) out of ${notifications.length} detected`);
                this.emit([latestBadge]);
            }

            this.lastPollTimestamp = Date.now();
            console.log('[BadgePolling] Poll complete');
        } catch (error) {
            console.error('[BadgePolling] Error during poll:', error);
        } finally {
            this.isPolling = false;
        }
    }

    // Force a refresh - useful when user manually refreshes badges
    async forceRefresh() {
        console.log('[BadgePolling] Force refresh triggered');
        await this.poll();
    }

    // Clear known badges (useful for testing or reset)
    clearKnownBadges() {
        localStorage.removeItem(KNOWN_BADGES_KEY);
        localStorage.removeItem(NOTIFIED_AVAILABLE_KEY);
        console.log('[BadgePolling] Cleared known badges cache');
    }

    // Get last poll timestamp
    getLastPollTimestamp(): number {
        return this.lastPollTimestamp;
    }
}

// Export singleton instance
export const badgePollingService = new BadgePollingService();

File: src-tauri/src/services/badge_polling_service.rs
use crate::commands::badges::get_cached_global_badges;
use crate::models::settings::AppState;
use crate::services::cache_service;
use crate::services::universal_cache_service::{get_cached_item, CacheType};

use chrono::{Datelike, Local, NaiveDate, TimeZone};
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::path::PathBuf;
use std::sync::Arc;
use tauri::{AppHandle, Emitter};
use tokio::sync::RwLock;
use tokio::time::{interval, Duration};

const POLL_INTERVAL_SECS: u64 = 5 * 60; // 5 minutes
const INITIAL_DELAY_SECS: u64 = 10;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BadgeNotificationStatus {
    New,
    Available,
    ComingSoon,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BadgeNotification {
    pub badge_name: String,
    pub badge_set_id: String,
    pub badge_version: String,
    pub badge_image_url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub badge_description: Option<String>,
    pub status: BadgeNotificationStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub date_info: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BadgeMetadata {
    pub date_added: Option<String>,
    pub usage_stats: Option<String>,
    pub more_info: Option<String>,
    pub info_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
struct BadgePollingStateFile {
    known_badges: Vec<String>,
    notified_available_badges: Vec<String>,
    last_poll_timestamp_ms: u64,
}

#[derive(Debug, Clone, Default)]
struct BadgePollingState {
    known_badges: HashSet<String>,
    notified_available_badges: HashSet<String>,
    last_poll_timestamp_ms: u64,
}

impl BadgePollingState {
    fn from_file(file: BadgePollingStateFile) -> Self {
        Self {
            known_badges: file.known_badges.into_iter().collect(),
            notified_available_badges: file.notified_available_badges.into_iter().collect(),
            last_poll_timestamp_ms: file.last_poll_timestamp_ms,
        }
    }

    fn to_file(&self) -> BadgePollingStateFile {
        BadgePollingStateFile {
            known_badges: self.known_badges.iter().cloned().collect(),
            notified_available_badges: self.notified_available_badges.iter().cloned().collect(),
            last_poll_timestamp_ms: self.last_poll_timestamp_ms,
        }
    }
}

pub struct BadgePollingService {
    running: Arc<RwLock<bool>>,
    is_polling: Arc<RwLock<bool>>,
    state: Arc<RwLock<BadgePollingState>>,
}

impl BadgePollingService {
    pub fn new() -> Self {
        Self {
            running: Arc::new(RwLock::new(false)),
            is_polling: Arc::new(RwLock::new(false)),
            state: Arc::new(RwLock::new(BadgePollingState::default())),
        }
    }

    pub async fn start(&self, app_handle: AppHandle, app_state: AppState) {
        // Check if already running
        {
            let mut running = self.running.write().await;
            if *running {
                return;
            }
            *running = true;
        }

        // Load persisted state once at startup
        if let Ok(state) = load_state_from_disk() {
            *self.state.write().await = state;
        }

        let running = self.running.clone();
        let is_polling = self.is_polling.clone();
        let state = self.state.clone();

        tokio::spawn(async move {
            // Let the app settle before first poll
            tokio::time::sleep(Duration::from_secs(INITIAL_DELAY_SECS)).await;

            // First poll
            if let Err(e) = poll_once(&app_handle, &app_state, &running, &is_polling, &state).await
            {
                eprintln!("[BadgePolling] Initial poll failed: {e}");
            }

            let mut ticker = interval(Duration::from_secs(POLL_INTERVAL_SECS));
            // Consume the immediate tick
            ticker.tick().await;

            loop {
                ticker.tick().await;

                // Stop if requested
                if !*running.read().await {
                    break;
                }

                // Avoid work if notifications are disabled
                let badge_notifications_enabled = {
                    let settings = app_state.settings.lock().unwrap();
                    settings.live_notifications.enabled
                        && settings.live_notifications.show_badge_notifications
                };

                if !badge_notifications_enabled {
                    continue;
                }

                if let Err(e) =
                    poll_once(&app_handle, &app_state, &running, &is_polling, &state).await
                {
                    eprintln!("[BadgePolling] Poll failed: {e}");
                }
            }

            println!("[BadgePolling] Service stopped");
        });
    }

    pub async fn stop(&self) {
        let mut running = self.running.write().await;
        *running = false;
    }
}

async fn poll_once(
    app_handle: &AppHandle,
    _app_state: &AppState,
    running: &Arc<RwLock<bool>>,
    is_polling: &Arc<RwLock<bool>>,
    state: &Arc<RwLock<BadgePollingState>>,
) -> anyhow::Result<()> {
    // Quick abort
    if !*running.read().await {
        return Ok(());
    }

    // Prevent overlapping polls
    {
        let mut polling = is_polling.write().await;
        if *polling {
            return Ok(());
        }
        *polling = true;
    }

    let result = poll_once_impl(app_handle, state).await;

    // Always clear polling flag
    *is_polling.write().await = false;

    result
}

async fn poll_once_impl(
    app_handle: &AppHandle,
    state: &Arc<RwLock<BadgePollingState>>,
) -> anyhow::Result<()> {
    // Get cached badges (already fetched by main.rs background prefetch)
    let cached_badges = match get_cached_global_badges().await {
        Ok(Some(b)) => b,
        Ok(None) => return Ok(()),
        Err(e) => return Err(anyhow::anyhow!(e)),
    };

    if cached_badges.data.is_empty() {
        return Ok(());
    }

    let mut notifications: Vec<BadgeNotification> = Vec::new();

    // We copy the state out for fast checks, then write back once at the end.
    let mut local_state = state.read().await.clone();

    for badge_set in &cached_badges.data {
        for version in &badge_set.versions {
            let badge_key = format!("{}-v{}", badge_set.set_id, version.id);

            let metadata_key = format!("metadata:{}", badge_key);
            let metadata = match get_cached_item(CacheType::Badge, &metadata_key).await {
                Ok(Some(entry)) => serde_json::from_value::<BadgeMetadata>(entry.data).ok(),
                _ => None,
            };

            let (status, date_info) = get_badge_status(metadata.as_ref());

            // New badge
            if !local_state.known_badges.contains(&badge_key) {
                local_state.known_badges.insert(badge_key.clone());

                let should_notify = matches!(
                    status,
                    Some(BadgeStatus::Available) | Some(BadgeStatus::ComingSoon)
                ) || status.is_none();

                if should_notify {
                    notifications.push(BadgeNotification {
                        badge_name: version.title.clone(),
                        badge_set_id: badge_set.set_id.clone(),
                        badge_version: version.id.clone(),
                        badge_image_url: first_badge_image_url(version),
                        badge_description: if version.description.is_empty() {
                            None
                        } else {
                            Some(version.description.clone())
                        },
                        status: BadgeNotificationStatus::New,
                        date_info: date_info.clone(),
                    });
                }
            }

            // Known badge that just became available
            if local_state.known_badges.contains(&badge_key)
                && status == Some(BadgeStatus::Available)
                && !local_state.notified_available_badges.contains(&badge_key)
            {
                local_state
                    .notified_available_badges
                    .insert(badge_key.clone());

                notifications.push(BadgeNotification {
                    badge_name: version.title.clone(),
                    badge_set_id: badge_set.set_id.clone(),
                    badge_version: version.id.clone(),
                    badge_image_url: first_badge_image_url(version),
                    badge_description: if version.description.is_empty() {
                        None
                    } else {
                        Some(version.description.clone())
                    },
                    status: BadgeNotificationStatus::Available,
                    date_info,
                });
            }
        }
    }

    // Persist updated state
    local_state.last_poll_timestamp_ms = current_time_ms();
    if let Err(e) = save_state_to_disk(&local_state) {
        eprintln!("[BadgePolling] Failed to persist state: {e}");
    }
    *state.write().await = local_state;

    // Emit only the latest badge notification (mirrors old TS behavior)
    if let Some(latest) = notifications.pop() {
        // Always emit the generic event
        let _ = app_handle.emit("badge-notification", vec![latest.clone()]);

        // Emit the specific event for availability transitions
        if matches!(latest.status, BadgeNotificationStatus::Available) {
            let _ = app_handle.emit("badge-available", vec![latest]);
        }
    }

    Ok(())
}

fn first_badge_image_url(version: &crate::commands::badges::HelixBadgeVersion) -> String {
    if !version.image_url_4x.is_empty() {
        return version.image_url_4x.clone();
    }
    if !version.image_url_2x.is_empty() {
        return version.image_url_2x.clone();
    }
    version.image_url_1x.clone()
}

fn current_time_ms() -> u64 {
    use std::time::{SystemTime, UNIX_EPOCH};

    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

fn state_file_path() -> anyhow::Result<PathBuf> {
    Ok(cache_service::get_app_data_dir()?.join("badge_polling_state.json"))
}

fn load_state_from_disk() -> anyhow::Result<BadgePollingState> {
    let path = state_file_path()?;

    if !path.exists() {
        return Ok(BadgePollingState::default());
    }

    let text = std::fs::read_to_string(&path)?;
    let file: BadgePollingStateFile = serde_json::from_str(&text)?;
    Ok(BadgePollingState::from_file(file))
}

fn save_state_to_disk(state: &BadgePollingState) -> anyhow::Result<()> {
    let path = state_file_path()?;

    let file = state.to_file();
    let text = serde_json::to_string_pretty(&file)?;
    std::fs::write(path, text)?;
    Ok(())
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum BadgeStatus {
    Available,
    ComingSoon,
    Expired,
}

fn get_badge_status(metadata: Option<&BadgeMetadata>) -> (Option<BadgeStatus>, Option<String>) {
    let Some(metadata) = metadata else {
        return (None, None);
    };

    let Some(more_info) = metadata.more_info.as_deref() else {
        return (None, None);
    };

    let Some((start, end, date_info)) = parse_date_range(more_info) else {
        return (None, None);
    };

    let now = Local::now();

    if now < start {
        return (Some(BadgeStatus::ComingSoon), Some(date_info));
    }

    if now >= start && now <= end {
        return (Some(BadgeStatus::Available), Some(date_info));
    }

    (Some(BadgeStatus::Expired), Some(date_info))
}

/// Parse a date range embedded in `more_info`.
///
/// Supported formats (matching the old TS implementation):
/// - "Dec 06 – Dec 07" / "Dec 06 - Dec 07"
/// - "Dec 1-12" / "Dec 1 - 12"
fn parse_date_range(
    text: &str,
) -> Option<(chrono::DateTime<Local>, chrono::DateTime<Local>, String)> {
    static FULL_RANGE: once_cell::sync::Lazy<Regex> = once_cell::sync::Lazy::new(|| {
        Regex::new(
            r"(?P<sm>[A-Za-z]{3})\s+(?P<sd>\d{1,2})\s*[–-]\s*(?P<em>[A-Za-z]{3})\s+(?P<ed>\d{1,2})",
        )
        .unwrap()
    });

    static SHORT_RANGE: once_cell::sync::Lazy<Regex> = once_cell::sync::Lazy::new(|| {
        // Negative lookahead (like TS) isn’t supported in Rust regex.
        // We keep it simple: match "Mon D-D" and allow extra trailing text.
        Regex::new(r"(?P<m>[A-Za-z]{3})\s+(?P<sd>\d{1,2})\s*[–-]\s*(?P<ed>\d{1,2})").unwrap()
    });

    static DATE_INFO: once_cell::sync::Lazy<Regex> = once_cell::sync::Lazy::new(|| {
        Regex::new(r"(?P<info>[A-Za-z]{3}\s+\d{1,2}(?:\s*[–-]\s*(?:[A-Za-z]{3}\s+)?\d{1,2})?)")
            .unwrap()
    });

    let year = Local::now().year();

    let date_info = DATE_INFO
        .captures(text)
        .and_then(|c| c.name("info").map(|m| m.as_str().to_string()))
        .unwrap_or_else(|| text.to_string());

    // Helper: month abbrev -> month number
    let month_num = |abbr: &str| -> Option<u32> {
        match abbr {
            "Jan" => Some(1),
            "Feb" => Some(2),
            "Mar" => Some(3),
            "Apr" => Some(4),
            "May" => Some(5),
            "Jun" => Some(6),
            "Jul" => Some(7),
            "Aug" => Some(8),
            "Sep" => Some(9),
            "Oct" => Some(10),
            "Nov" => Some(11),
            "Dec" => Some(12),
            _ => None,
        }
    };

    // Full month-to-month range
    if let Some(caps) = FULL_RANGE.captures(text) {
        let sm = month_num(caps.name("sm")?.as_str())?;
        let sd: u32 = caps.name("sd")?.as_str().parse().ok()?;
        let em = month_num(caps.name("em")?.as_str())?;
        let ed: u32 = caps.name("ed")?.as_str().parse().ok()?;

        let start_date = NaiveDate::from_ymd_opt(year, sm, sd)?;
        let end_date = NaiveDate::from_ymd_opt(year, em, ed)?;

        let start = Local
            .from_local_datetime(&start_date.and_hms_opt(0, 0, 0)?)
            .single()?;
        let end = Local
            .from_local_datetime(&end_date.and_hms_opt(23, 59, 59)?)
            .single()?;

        return Some((start, end, date_info));
    }

    // Short day range within same month
    if let Some(caps) = SHORT_RANGE.captures(text) {
        let m = month_num(caps.name("m")?.as_str())?;
        let sd: u32 = caps.name("sd")?.as_str().parse().ok()?;
        let ed: u32 = caps.name("ed")?.as_str().parse().ok()?;

        let start_date = NaiveDate::from_ymd_opt(year, m, sd)?;
        let end_date = NaiveDate::from_ymd_opt(year, m, ed)?;

        let start = Local
            .from_local_datetime(&start_date.and_hms_opt(0, 0, 0)?)
            .single()?;
        let end = Local
            .from_local_datetime(&end_date.and_hms_opt(23, 59, 59)?)
            .single()?;

        return Some((start, end, date_info));
    }

    None
}
